test/Constraints/static_members_on_protocol_metatype.swift

```diff
@@ -0,0 +1,215 @@
+let _: S = P[42] // Ok
+let _: S = P[42].other // OK
+let _: G<Int> = P[t: 42] // Ok
+let _: G = P[t: 42] // Ok
+let _: G<String> = P[t: 42] // expected-error {{cannot convert value of type 'Int' to expected argument type 'String'}}
+let _: G<Int> = P[t: 42].other // Ok
+let _: G<String> = P[t: 42].other // expected-error {{cannot convert value of type 'Int' to expected argument type 'String'}}
+
+func test<T: P>(_: T) {}
+
+test(.property) // Ok, base is inferred as Style.Type
+test(.property.other) // Ok
+test(.fnProp()) // Ok
+test(.fnProp().other) // Ok
+test(.method()) // Ok, static method call on the metatype
+test(.method().other) // Ok
+test(.genericFn(42)) // Ok
+test(.genericFn(42).other) // Ok
+
+protocol Q {}
+
+func test_combo<T: P & Q>(_: T) {} // expected-note 2 {{where 'T' = 'G<Int>'}}
+
+extension Q {
+  static var otherProperty: S { S() }
+
+  static func otherMethod() -> S {
+    return S()
+  }
+
+  static func otherGeneric<T>(_: T) -> S {
+    return S()
+  }
+}
+
+extension S : Q {
+}
+
+test_combo(.property) // Ok
+test_combo(.method()) // Ok
+test_combo(.otherProperty) // Ok
+test_combo(.otherProperty.other) // Ok
+test_combo(.otherProperty.property) // expected-error {{static member 'property' cannot be used on instance of type 'S'}}
+test_combo(.otherMethod()) // Ok
+test_combo(.otherMethod().method()) // expected-error {{static member 'method' cannot be used on instance of type 'S'}}
+test_combo(.otherGeneric(42)) // Ok
+
+test_combo(.genericFn(42)) // expected-error {{global function 'test_combo' requires that 'G<Int>' conform to 'Q'}}
+
+/* Invalid result types */
+
+extension P {
+  static var invalidProp: Int { 42 } // expected-note 5 {{'invalidProp' declared here}}
+  static var selfProp: Self { fatalError() }
+  static func invalidMethod() -> Int { 42 } // expected-note 5 {{'invalidMethod()' declared here}}
+  static func generic<T>(_: T) -> T { fatalError() } // expected-note 5 {{'generic' declared here}}
+  static func genericWithReqs<T: Collection, Q>(_: T) -> Q where T.Element == Q { // expected-note 3 {{'genericWithReqs' declared here}} expected-note 3 {{required by static method 'genericWithReqs' where 'T' = '()'}}
+    fatalError()
+  }
+
+  static subscript(q q: String) -> Int { get { 42 } } // expected-note 2 {{'subscript(q:)' declared here}}
+}
+
+_ = P.doesntExist // expected-error {{type 'P' has no member 'doesntExist'}}
+_ = P.selfProp // expected-error {{generic parameter 'Self' could not be inferred}}
+_ = P.invalidProp
+// expected-error@-1 {{cannot reference static property 'invalidProp' on 'P.Protocol' with non-conforming result type 'Int'}}
+_ = P.invalidProp.other
+// expected-error@-1 {{cannot reference static property 'invalidProp' on 'P.Protocol' with non-conforming result type 'Int'}}
+// expected-error@-2 {{value of type 'Int' has no member 'other'}}
+_ = P.invalidMethod()
```

Comment:
@xedin Just for curiosity, is possible for one to try reference the method like `_ = P.invalidMethod`? If it is what happens in this case?

Reply:
It's possible to get a function by value if its result type conform to a required protocol because it would just mean partial application, looks like I didn't add any tests for that, although I should have...

Reply:
>  it would just mean partial application

Ah got it, thanks :)